### PR TITLE
fix(net): bound and stop re-sorting getqrinfo base block hashes

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
## Issue being fixed or feature implemented

`CGetQuorumRotationInfo::baseBlockHashes` is deserialized with no size limit, and `GETQUORUMROTATIONINFO` is served to any peer, with no rate limiting, while holding `cs_main`. Two things compound:

1. The wire format accepts `MAX_PROTOCOL_MESSAGE_LENGTH / sizeof(uint256)` = 98304 base block hashes in a single 3 MiB message. Each one costs a block-index lookup and an active-chain check under `cs_main`.
2. On the non-legacy construction path `GetLastBaseBlockHash()` sorts the *entire* base list on every call, and `BuildQuorumRotationInfo()` calls it once per constructed `CSimplifiedMNListDiff`. For `llmq_60_75` that is up to `3 * signingActiveQuorumCount` (96) snapshot bases plus the target cycles and the tip, so roughly 10-30 full sorts per request. Every comparison dereferences a `CBlockIndex*`, so this is cache-hostile work that scales as `k * n log n` with `n` chosen by the requesting peer.

The requesting peer does have to supply genuine hashes, since the population loop returns early on the first hash that is not found or is not in the active chain, but on a chain with millions of blocks that costs an attacker nothing, and the peer picks the construction path by advertising its protocol version.

Found while re-verifying an audit finding against `develop`; there is no open issue.

## What was done

Two independent changes, one commit each.

**Bound the request.** `baseBlockHashes` now deserializes through `LIMITED_VECTOR(..., MAX_BASE_BLOCK_HASHES)` with `MAX_BASE_BLOCK_HASHES = 4096`. `LimitedVectorFormatter` emits the ordinary vector wire format, so nothing changes for senders. Only the highest base at or below each diff target is ever used, and a response builds at most ~101 diffs for `llmq_60_75`, so no request can usefully carry more than that. I checked what shipping clients actually send:

| Client | Bases per request | Where |
|---|---|---|
| DashSync (iOS) | 1 | `DSQuorumRotationService.m`, `-getQRInfoForBlockHash:previousBlockHash:` |
| dashj (Android) | at most 6, deduplicated | `QuorumRotationState.java`, `getQuorumRotationInfoRequest` |
| rust-dashcore `dash-spv` | 0 or 1 | `sync/masternodes/manager.rs`, `send_qrinfo_for_tip` |
| Platform | does not send `getqrinfo` | |

So 4096 is roughly 40x the theoretical maximum useful list and ~700x real usage, while cutting the attacker's worst case 24x.

The handler now catches the deserialization failure locally and scores the peer with `Misbehaving(100)`, the same shape `FILTERLOAD`, `FILTERADD` and `SPORK` already use, instead of letting the outer `ProcessMessages()` catch drop the message with only a `BCLog::NET` line. An honest client that somehow exceeded the cap would otherwise hang with no signal on either side. `BuildQuorumRotationInfo()` also checks the cap itself so `quorum rotationinfo`, which builds the request in memory without a serialization round-trip, returns an RPC error instead of doing unbounded work.

**Stop re-sorting.** `GetLastBaseBlockHash()` no longer sorts; sortedness is now a documented precondition, and it takes `Span<const CBlockIndex* const>` so it cannot mutate the caller's list. The two callers that appended mid-construction now use `InsertBaseBlockSorted()`, which inserts at `std::upper_bound` instead of `push_back`, preserving the ordering at `O(n)` pointer moves rather than re-establishing it at `O(n log n)` index dereferences. With the list guaranteed sorted, the lookup itself is a binary search. The initial `std::sort` in `BuildQuorumRotationInfo()` is unchanged and still runs for both construction paths. Both the sort and the insert use `node::CBlockIndexHeightOnlyComparator` so one definition owns the ordering, and the genesis fallback reads the precomputed consensus hash instead of re-hashing the genesis header per call.

Output is unchanged. The legacy path never sorted inside `GetLastBaseBlockHash()` and never appends, so it is untouched. On the non-legacy path the list was already sorted and deduplicated before the first call, and every subsequent append was followed by a re-sort; inserting in position produces the same sequence. Equal heights can only mean the same active-chain block, so ties resolve to the same hash either way.

This PR removes a secondary amplifier. The dominant cost of the handler, ~30 `BuildSimplifiedMNListDiff` calls under `cs_main` with no per-peer rate limit, is unchanged and is a separate piece of work.

## How Has This Been Tested?

Built with `--enable-debug --enable-crash-hooks` on aarch64-apple-darwin.

Unit tests, `src/test/llmq_snapshot_tests.cpp`:

- `get_quorum_rotation_info_base_block_hashes_limit_test` is new: a request at exactly `MAX_BASE_BLOCK_HASHES` round-trips, and one past it throws `std::ios_base::failure` with the target vector still empty and the element bytes still unread, which is the "rejected before any element is decoded" property the DoS argument rests on.
- `get_last_base_block_hash_repeated_base_blocks_test` was updated for the new contract: the case that fed deliberately unsorted input to exercise the internal sort is gone, and cases covering the genesis fallback for an empty list and for a list entirely above the target were added.

Functional test, `test/functional/feature_llmq_rotation.py`:

- The existing `quorum rotationinfo` block now also asserts that two base blocks at different heights produce identical output regardless of request order, and that the second base is not inert: it displaces the genesis fallback in `mnListDiffAtHMinus3C`.
- New `test_getqrinfo_base_block_hashes_limit` drives `getqrinfo` over a real P2P connection using a new `msg_getqrinfo` framework message. A request at the limit is answered with `qrinfo` and no `Misbehaving`; one past the limit, and a bare CompactSize prefix with no element bytes at all, both log `Misbehaving` with `malformed getqrinfo received` and disconnect the peer. The RPC path with the same oversized list returns `too many baseBlockHashes`.

`llmq_snapshot_tests`, `coinjoin_inouts_tests`, `net_tests` and `feature_llmq_rotation.py` pass locally. `lint-python.py`, `lint-whitespace.py`, `lint-circular-dependencies.py`, `lint-includes.py` and `clang-format-diff` are clean on the diff.

## Breaking Changes

None for any current client. A `getqrinfo` carrying more than 4096 base block hashes is now rejected and the peer is scored as misbehaving; no known client sends more than six, and Dash Core itself never sends this message. Serialization is byte-identical, so senders are unaffected. `doc/release-notes-7629.md` documents the new limit.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by thepastaclaw** [comment]
✅ Final review complete — no blockers (commit dea5c4be) · triage: critical · Phase 2 only (queue backlog)

**Review by thepastaclaw — APPROVED** [review]
## Final validation — Codex/Sol only (Phase 2 disabled)

The exact-head diff correctly bounds unauthenticated getqrinfo deserialization while retaining the existing wire encoding, and it preserves the sorted base-block invariant without repeated full-list sorts. Call-site and test inspection found no actionable correctness, compatibility, performance, or coverage issues.
Source: reviewers codex-general and codex-dash-core-commit-history (exact backend model IDs were not included in the supplied evidence); final verifier Claude (exact runtime model ID was not exposed). openclaw-agent/cliproxy/gpt-5.6-sol is orchestration-only and is not reviewer evidence.

Validated zero-blocker Codex/Sol precheck evidence was promoted to final because Phase 2 (Sonnet/Opus) is temporarily disabled. This is Codex/Sol-only final validation, not Codex + Sonnet/Opus coverage.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — dash-core-commit-history (completed)
- Verifier: `gpt-5.6-sol` — verifier
- Sonnet/Opus: **not run (Phase 2 disabled — temporary Codex/Sol-only final)**
- Secondary pass: disabled (`temporary_phase2_sonnet_disable`)

File: src/llmq/snapshot.cpp [hunk 1]
```diff
@@ -219,45 +234,36 @@ bool BuildQuorumRotationInfo(CDeterministicMNManager& dmnman, CQuorumSnapshotMan
...
-                  [](const CBlockIndex* a, const CBlockIndex* b) { return a->nHeight < b->nHeight; });
-    }
-    // default to genesis block
-    uint256 hash{Params().GenesisBlock().GetHash()};
-    for (const auto baseBlock : baseBlockIndexes) {
-        if (baseBlock->nHeight > blockIndex->nHeight) break;
-        hash = baseBlock->GetBlockHash();
-    }
-    return hash;
+    Assume(std::is_sorted(baseBlockIndexes.begin(), baseBlockIndexes.end(), CBlockIndexHeightOnlyComparator()));
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/llmq/snapshot.cpp]: Good catch. `Assume` evaluates its argument in release builds, so the `is_sorted` check reintroduced a linear scan per lookup. Removed in dea5c4bea0; the sorted precondition is enforced by construction (initial `std::sort` plus `InsertBaseBlockSorted`) and stated in the header comment.

---
🤖 Posted autonomously by Codex on behalf of pasta.

**Review by thepastaclaw — APPROVED** [review]
## Final validation — Phase 2 only (queue backlog)

No actionable in-scope defects were found at dea5c4bea09b5f40ec5e15ef3fc741aeaae53bd4. Source inspection confirms rejection before vector-element decoding, the shared RPC construction limit, and preservation of sortedness for binary-search lookups; the earlier thread's per-lookup sortedness scan is absent. The two commits are narrowly scoped and include relevant regression tests; diff whitespace checks passed and the worktree is clean, but the reported build and test executions were not independently rerun.

Source: reviewer 1: `gpt-6-astra` (agent: `phase2-reviewer`, role: `general`); reviewer 2: `gpt-6-astra` (agent: `phase2-reviewer`, role: `dash-core-commit-history`); final verifier: `gpt-6-astra` (agent: `astra-verifier`, role: `final-verifier`)

### Review provenance
- Triage: `critical` by `gpt-6-astra` (effort low) — This changes peer-controlled network deserialization, peer punishment, and quorum-rotation response construction under cs_main, where mistakes could cause denial of service, disconnect legitimate peers, or return incorrect quorum synchronization data.
- Phase 1 reviewers: **not run (skipped for throughput: 21 PRs queued, above the 10 limit)**
- Fresh verifier: `gpt-6-astra` — final-verifier; agent `astra-verifier`
- Phase 2 reviewers: `gpt-6-astra` — general (completed, effort xhigh); agent `phase2-reviewer`, `gpt-6-astra` — dash-core-commit-history (completed, effort xhigh); agent `phase2-reviewer`
